Don't throw an error when registering a payment method fails

Aljullu opened the pull request:
Fixes #3125.
Fixes #3157.

This PR has some changes in how failing payment methods are handled: they will silently fail for shoppers, but a notice will be shown for admins.

### How to test the changes in this Pull Request:

Excuses in advance for the long testing steps, this PR touches some logic that affects several critical flows so I tried to list the steps as detailed as possible.

### Test #3125.
##### Preparation
Steps from #3125:
1. Install & activate WooCommerce Stripe. 
2. Enable Stripe CC payment method - don't add an api key (or delete the option).
3. Enable BACS or other payment methods (e.g. cheque).
3. Add checkout block to checkout page.
4. On front end, add something to cart and proceed to checkout.

##### Tests
1. Make sure you have an admin user with saved payment methods<sup>*</sup> and a user without saved payment methods (or a guest).
2. With those two users, check the Checkout page in the frontend. Also open the Checkout page in the editor.
3. Verify the statements below are true. :point_down: 

| Scenario | With Stripe API key | Without Stripe API key |
| --- | --- | --- |
| Admin with saved payment methods / Editor | <ul><li>Saved payment methods are displayed.</li><li>Credit Card payment method is displayed (under use a new payment method).</li><li>There is no error notice.</li></ul> | <ul><li>Saved payment methods are not displayed.</li><li>There is an error notice about Stripe not being registered correctly.</li><li>Other payment methods are displayed as usual.</li></ul> |
| Guest user | <ul><li>Saved payment methods are not displayed.</li><li>Credit Card payment method is displayed.</li><li>There is no error notice.</li></ul> | <ul><li>Saved payment methods are not displayed.</li><li>Credit Card payment method is not displayed.</li><li>There is no error notice.</li><li>Other payment methods are displayed as usual.</li></ul> |

<sup>*</sup> In order to save a payment method with a user. Enable the WooCommerce Stripe plugin, set the keys and make a purchase with a user selecting the `Save payment information to my account for future purchases.` option. Next time you visit the Checkout with that user, the saved payment method will show up.

#### Test #3157
Assuming you already have a user with saved credit cards in Stripe from the steps above.
1. Go to Stripe settings and uncheck `Enable Payment via Saved Cards`. Make sure you have added back the API keys that you might have removed in the steps above.
2. Start a purchase with a user that has saved payment methods and go to the Checkout block.
3. Verify saved credit cards are not shown in the Payment method options.

#### Regression testing

It would also be great to do a purchase flow with some of the scenarios described above to ensure the payment flow still works properly.

### Changelog

> Fixed an issue that was not showing other payment methods if one failed to load when the user was an admin.

haszari commented on assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js:
```diff
@@ -24,7 +24,9 @@ import {
-		throw new Error(
+		// This is only shown to admin users, so it's ok to print an error to the browser console.
+		// eslint-disable-next-line no-console
+		console.error(
```
Seems like we have a few lines of code just to throw up a `console.error` – is it worth it? 

[Later in the file, there's a comment about this info being presented as a notice.](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/5c5bd3a70c7d96208c5f975cfc2864f9d665d6c7/assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js#L126) That makes sense to me – if we're going to add explicit support for admins of misconfigured sites, we might as well use notice to show in the UI.

haszari reviewed (commented):
Switching to `console.error` fixes the issue, but I'm thinking we should revisit our intentions here, and either make this a real feature, or remove this special support for admins.

There are two options as I see it:

1. Silently hide the misconfigured payment method – i.e. same behaviour for admin and guest/customer.  There are other ways for the merchant to figure out why their payment method is not working - this isn't a core responsibility of the block front-end code.
2. Show a notice (to admins) informing them that there's a problem with one of their payment methods. 

Option 2 seems preferable, based on the comments in the code, seems like it's what we intended/designed.

Option 1 seems like a fine fallback if we hit issues/complexity with other options. While this could help some merchants, it doesn't feel like an essential feature to me. Having said that, I'm not clear on our approach/policy for showing "editor"/merchant notices or guidance in the front end.

Note – if for some reason it's best to go with shipping this as a `console.error`, I don't see any issue with that (i.e. all the above is non blocking). But in that case please tweak the comments to make the intention clear 🔮

Aljullu commented:
Thanks for the review @haszari!

> Switching to `console.error` fixes the issue, but I'm thinking we should revisit our intentions here, and either make this a real feature, or remove this special support for admins.

I thought it was the expected behavior to show a console error instead of a frontend notice when a payment method fails to load in the frontend, but tagging @nerrad for confirmation.

In 8985a56337b209cf7af63f52f12e6876f602b89b I updated this PR to use notices instead of a console error and I also updated the PR testing steps. We can easily revert these changes depending on the final decision taken.

nerrad commented:
The intention with the thrown error was that when admins are logged in, the error boundary would catch the error. At one point I think it did, but somewhere along the way it stopped. So ideally I'd like to some investigation around why the error boundary isn't working to catch this error (we might have to implement the `useThrowError` somewhere here (I haven't looked closely at the code).

Aljullu commented:
Ah yes, the error boundary is used in the editor to display the error. I will move this PR back to in progress so I can investigate it further.

Aljullu commented:
Hmmm, I have been investigating this and I'm not sure if it ever worked as described. If it did, I couldn't find when/how. The issue seems to be that we are throwing inside a `for` loop, so the execution stops and other payment methods are not loaded. To fix that we would need to directly throw on payment method initialization inside the component that renders it. This is what I tried to do in the last commits in this PR.

What they do is to enable all payment methods for admin users, so if one is failing, the error will be caught by the error boundary and shown as a notice. That works mostly fine, but I needed to add a filter to hide saved payment options for methods that had thrown an error, because by default they are unfiltered.

I'm not a big fan of this approach since it requires some special logic for a specific use case (admin user), but I can't think of any better approach (ideas welcome).

I split the commits in self-contained changes, so it might be easier to review commit-by-commit than all at once. I also updated the testing instructions in the PR description.

haszari commented:
> Hmmm, I have been investigating this and I'm not sure if it ever worked as described. If it did, I couldn't find when/how. The issue seems to be that we are throwing inside a for loop, so the execution stops and other payment methods are not loaded. To fix that we would need to directly throw on payment method initialization inside the component that renders it. This is what I tried to do in the last commits in this PR.
> What they do is to enable all payment methods for admin users, so if one is failing, the error will be caught by the error boundary and shown as a notice. That works mostly fine, but I needed to add a filter to hide saved payment options for methods that had thrown an error, because by default they are unfiltered.
> I'm not a big fan of this approach since it requires some special logic for a specific use case (admin user), but I can't think of any better approach (ideas welcome).

Valiant effort @Aljullu !! 🙌😅

If this feature hasn't ever worked as intended, maybe it's better to keep this simple, and remove the feature (i.e. admins get the exact same behaviour as other users). This is not an introduced issue, and the impact is small – this is a small convenience to give admins a clue in the front end why Stripe is not visible (or another misconfigured payment gateway). We can always implement it later if there's a clear need for it.

I haven't reviewed the recent code changes – I think we should clarify that we actually want/need this feature before investing any further time. 

@nerrad do you have thoughts on this – do we need to display an admin-only notice if a payment method fails init or is misconfigured?

> What they do is to enable all payment methods for admin users, so if one is failing, the error will be caught by the error boundary and shown as a notice.

My naive assumption when I heard about this feature was that there'd be a try/catch within the loop body, to protect against rogue payment methods. Something like this:

```js
// pseudocode!!!
// initialising payment methods and deciding whether they should be shown as tabs
foreach ( current in all_payment_methods ) {
  try {
    current.init();
    if ( current.canMakePayment() ) { 
       addAvailablePaymentMethod( current );
    }
  catch ( error ) {
     if ( userIsAdmin ) {
        addNotice( { message: `Payment method ${ current.name } failed initialisation with error ${ error }` } );
     } 
  }
}
```

However, personally I don't think this is worth following up now, and there might be extra complexity getting the error handling right across components/hooks in a react context.

Aljullu commented:
Thanks for the feedback @haszari! Yeah, let's wait for @nerrad's thoughts in order to decide how to proceed.

> My naive assumption when I heard about this feature was that there'd be a try/catch within the loop body, to protect against rogue payment methods. 

There isn't but in some way [React error boundaries](https://reactjs.org/docs/error-boundaries.html) work similarly (but in a React-ish way). So if an error happens inside a component, the error boundary catches it and handles it as required; in this case, showing a notice.

Here, the error is thrown in the payment method components and caught by `PaymentMethodErrorBoundary`.

nerrad commented:
> @nerrad do you have thoughts on this – do we need to display an admin-only notice if a payment method fails init or is misconfigured?

Short answer yes.

Long answer:  Anything we can do to make it easier for merchant's to more easily see what's wrong (especially when there are potential configuration issues) is of benefit. As for why admin only, I don't think we want to expose faulty payment method options to shoppers or potentially expose sensitive info to shoppers given we have no control over the contents of error notices. So restricting display of these notices to only Admin's limits the potential problematic surface area.

As for this feature never working, that may be the case, but the intention was still always to display useful error messages to admins of a store for troubleshooting configuration issues.

@Aljullu, assuming this tests well (I haven't tested) I'm fine with rolling with this iteration for now. However, I've pulled from the 3.4.0 milestone to give it a bit more time on the main branch for our development cycle (which hopefully will surface any potential issues between releases).

haszari commented on assets/js/base/components/payment-methods/payment-method-error-boundary.js:
```diff
@@ -36,11 +36,14 @@ class PaymentMethodErrorBoundary extends Component {
 					);
 				}
 			}
-			return (
```
Comment on the init of `let errorText =` in existing code:

```js
let errorText = __(
				'This site is experiencing difficulties with this payment method. Please contact the owner of the site for assistance.',
				'woo-gutenberg-products-block'
			);
```

Do we need the default/front end error message anymore? I'm not seeing it in the front end as shopper, so I'm assuming it's not used. 

Or to clarify the intent of this error boundary object .. it catches errors from individual payment methods, and if user is admin, shows an error notice inline about the problem with the payment method. 

Another comment on the existing code - `isEditor || CURRENT_USER_IS_ADMIN` - can we simplify this check?

Aljullu replied:
> Do we need the default/front end error message anymore? I'm not seeing it in the front end as shopper, so I'm assuming it's not used.

I think so. The error boundary can catch _any_ error thrown in the inner components, that means there might be unexpected errors that we can't anticipate with 3rd-party payment methods. The default error message allows the block to display a friendly message for shoppers.

> Another comment on the existing code - `isEditor || CURRENT_USER_IS_ADMIN` - can we simplify this check? 

What do you mean by simplify? Non-admins can also see the block in the editor, so in practice `isEditor` and `CURRENT_USER_IS_ADMIN` are different checks. I think it makes sense explicitly using both, but maybe I'm missing something? :thinking:

haszari replied:
Thanks for explaining about the default error message. I might test throwing a generic unexpected error to confirm the behaviour.

With the logic, I agree it probably doesn't need changing, that check is complete. I think I was worried about the "`isEditor` but can't change payment method settings" case.

Aljullu replied:
You mean a user that can see the block in the editor but doesn't have permissions to modify the payment methods settings? That's a good point, it might make sense having a separate error message for those cases. I would keep it for a follow-up, however, so we don't make this PR more complex than what already is. :slightly_smiling_face:

haszari reviewed (changes requested):
I found a bug with this - haven't yet done a full test/review yet sorry. Lots of code impact here so will need to review carefully!

The issue I found is that if COD is dependent on shipping method, it's always shown. It's supposed to hide (dynamically) if a non-COD-compatible shipping method is selected. I switched branch to `main` to confirm the correct behaviour – so confirming that the issue is introduced with these changes. It's a delicate balance getting these payment tabs working well.

Here's the relevant option in COD config: 

<img width="1305" alt="Screen Shot 2020-09-15 at 10 03 26 AM" src="https://user-images.githubusercontent.com/4167300/93142788-d14efd00-f73a-11ea-8a17-01857c36ea00.png">

Aljullu commented:
> The issue I found is that if COD is dependent on shipping method, it's always shown. It's supposed to hide (dynamically) if a non-COD-compatible shipping method is selected. I switched branch to main to confirm the correct behaviour – so confirming that the issue is introduced with these changes. It's a delicate balance getting these payment tabs working well.

Aaah, good catch! That was a consequence of loading all payment methods for admins, without checking what was returned by `canMakePayment`. In 5856bb86c4a3063eebcd9044518930d56a53efb0 I changed this behavior so payment methods that don't allow payments aren't shown.

haszari reviewed (commented):
Epic work on this @Aljullu - this is a really tricky issue and you're really getting to the heart of it. 🎖️ 🦖 

I found another issue (sorry). And also note I've paused reviewing/testing because of the issue I hit - so this is not a full review. 

Here's how to reproduce the issue:

- Enable COD and/or Cheque payment method.
- Add a `throw` statement in `canMakePayment` for COD or Cheque payment method.
- View checkout block in front end.

Expected: see all valid payment methods as tabs, any saved cards as radio buttons, and a tab for the broken payment method (COD or cheque) with a notice.

Actual: no payment methods! Note this applies to admin and anon shopper.

Now .. why was I testing this? The most important part of this issue is that we want to ensure payment which throw errors are not able to break the checkout (i.e. other payment options should be unaffected). 

I also saw another specific issue with Stripe (kind of a separate issue but related). If there's no key, it returns an api object with an `error` property, but our check for successful init is if the api object is truthy. We probably need to check for `stripe.error` and if so, throw the error for the notice.

```js
	return canPayStripePromise.then( ( stripe ) => {
                // Need to check `stripe.error` here and if so throw the error for the notice.
		if ( stripe === null ) {
			isStripeInitialized = true;
			return canPay;
		}
		// Do a test payment to confirm if payment method is available.
                // This line is an error if API keys are missing, because stripe.paymentRequest is not a function.
		const paymentRequest = stripe.paymentRequest( {
			total: {
				label: 'Test total',
				amount: 1000,
			},
			country: getSetting( 'baseLocation', {} )?.country,
			currency: currencyCode,
		} );
		return paymentRequest.canMakePayment().then( ( result ) => {
			canPay = !! result;
			isStripeInitialized = true;
			return canPay;
		} );
	} );
```

Aljullu commented on assets/js/base/components/payment-methods/saved-payment-method-options.js:
```diff
@@ -92,63 +93,53 @@ const SavedPaymentMethodOptions = ( { onSelect } ) => {
+				return typeMethods.map( ( paymentMethod ) => {
+					const method =
+						standardMethods[ paymentMethod.method.gateway ];
+					if ( ! method?.supports?.savePaymentInfo ) {
```
This block of code might be easier to review with `Hide whitespace changes`. The only change inside this `useEffect` which is not a refactor is this new check: if `savePaymentInfo` is not truthy, the saved payment methods are not appended (ie: if Stripe fails to load, saved credit cards will not be appended).

Before, that would be achieved because the payment method would not have been registered at all. But with the changes in this PR, all payment methods are registered for admins, so we need to add this extra check.

Aljullu commented on assets/js/base/components/payment-methods/saved-payment-method-options.js:
```diff
@@ -167,7 +208,7 @@ const SavedPaymentMethodOptions = ( { onSelect } ) => {
 	// In the editor, show `Use a new payment method` option as selected.
 	const selectedOption = isEditor ? '0' : selectedToken + '';
-	return currentOptions.current.length > 0 ? (
+	return currentOptions.current.length > 1 ? (
```
That was part of the refactor in 73a19d1dbcee0f6d23cb17d1f826948dc461552a (see [related lines of code](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3134/files#diff-40dcd5cbcf22a46ff3ce6512834ba7f8L134-L144)) in order to make it easier to understand. Thinking about it twice, I should have broken it in two different PRs in order to make it easier to review. Let me know if you want me to split it.

The idea behind the refactor was to remove the `if` clause from the effect and the `push` to the array stored in the ref. I was having a hard time understanding how they worked and felt I was simplifying it. But I might have achieved the opposite. :sweat_smile: 

I did another try to simplify this (e3ca65d3fac7dfbb09a7003bea368177ce8cf523), basically instead of appending the default option inside the effect, we append it in the render.

Aljullu commented on assets/js/base/components/payment-methods/saved-payment-method-options.js:
```diff
@@ -92,63 +96,100 @@ const SavedPaymentMethodOptions = ( { onSelect } ) => {
+	const updateOptions = useCallback( async () => {
+		// Admin users have all payment methods enabled, so we need to filter
+		// the ones that do not accept payments.
+		const getFilteredPaymentMethods = async ( paymentMethods ) => {
```
All good points, I agree calling `canMakePayment` in two different places was not ideal. In fact that was what was making me not to be really comfortable with that approach either.

I took another take at this issue. Instead of filtering saved payment methods with `canMakePayment`, now I filter them with the `supports.savePaymentInfo` property, which we now set to false if the payment failed to register.

I also considered adding another attribute like `error` or `hasError` to 'mark' the payment methods that failed to register but are still added because the user is an admin. But I feel setting `savePaymentInfo` to false is a more natural approach.

In addition to that, I needed to replace the `content` so the error is thrown in the component (and then caught by the error boundary). Let me know if this new approach looks good to you.

Aljullu commented:
> I found another issue

Good catch! Thanks for testing and finding all those issue @haszari! The issue should be fixed now, that was because I removed the `try / catch` block around `canMakePayment`. I added it again in aa8ddda56e46ebd3d02cbc4cd948625826121c32.

Aljullu commented:
While working on this PR I noticed an issue in `main` (#3157). Fortunately, it's fixed in this PR with the new approach to check `savePaymentInfo` before saved payment methods are listed.

haszari reviewed (commented):
Nice work @Aljullu - the code is looking good now.

I tested extensively and (😢) found a few issues and quirks. I think only one of these is a potential blocker.

### If COD or cheque payment methods are "broken", purchase still completes. 
Note this is a contrived test situation, so this could be an edge case, but I think it's a signal that we need to make the API more robust, so rogue payment methods don't cause real issues for merchants.

![Screen Capture on 2020-09-17 at 15-49-49](https://user-images.githubusercontent.com/4167300/93418002-7a455580-f8fd-11ea-8e49-60308ea751b2.gif)

Here's how to reproduce. 

1. Enable COD.
2. "Break" COD in `canMakePayment()` by `throw`ing an error.
3. Add stuff to cart, proceed to checkout, select COD.
4. Click `Place order`.

I admit this is a contrived situation :) The issue I'm concerned about is that if any payment method has a bug (throws an exception) then it should not be possible to complete purchase with it. I suspect this is only an issue for admin, so it's low-impact, but might be important to get right.

### Other issues
These are all pretty minor – we could _probably_ ship with these. At the same time I think these could be useful feedback on the payment methods API / implementation which might lead to changes.

The error notice for Stripe express methods is different (more raw!) than the notice shown in Stripe credit card tab. I think I mentioned this in a previous review, it can probably be handled as a follow up issue.

<img width="1020" alt="Screen Shot 2020-09-17 at 2 14 41 PM" src="https://user-images.githubusercontent.com/4167300/93418433-81b92e80-f8fe-11ea-88ad-52c995225ec5.png">
<img width="1047" alt="Screen Shot 2020-09-17 at 2 14 35 PM" src="https://user-images.githubusercontent.com/4167300/93418435-854cb580-f8fe-11ea-9774-c41fe13316b7.png">

In the editor, the error notices aren't always visible, because of the tabbed UI, and the fact that it's static (editor can't interact with it). This is definitely a quirk, and limits the usefulness of generating an error notice; probably it's most important that it's displayed in the editor (if anywhere). To reproduce: break a payment method that is not the first tab (probably other ways to reproduce).

Also in the editor, the cheque error notice is simply not shown at all. This is curious – and suggests that the mechanism for showing error notices has a loophole somewhere. Again, this is a contrived situation (`throw` in `canMakePayment`), so might be limited impact.

### Expectations of the payment method API/components

I don't think it's very fair of me to just report all these detailed quirky issues! 

So I thought I'd clarify what my expectations are for our payment methods layer. (Note - this is personal opinion; it might be useful for us to discuss this as a team and document it as part of the API "contract" to payment gateways and merchants.)

When I'm testing, I keep these principles in mind:

- If a payment gateway has an error/bug or is misconfigured: 
  - The payment method cannot be used to complete checkout (ever!).
  - For shoppers: it is not shown. 
  - For merchant (admin, editor etc): a notice is shown to inform the merchant of the issue (so they can fix it!).
- If the gateway supports saved cards (aka saved payment methods) these are only shown when the gateway is functioning correctly.
- No "rogue" payment method can break the checkout or affect other payment methods.
- Editor should be roughly representative of the front end – i.e. same payment methods visible, same admin error notices, etc.

Aljullu commented:
Thanks for your patience @haszari, I know this PR is not easy to review! I completely changed the approach I was using, but I will answer individually the issues you reported:

> If COD or cheque payment methods are "broken", purchase still completes.

That's a big issue, I shouldn't have missed it! :facepalm: Given this and the other issues you described below, I think the approach to always register the payment methods for admin users is probably not the way to go.

> The error notice for Stripe express methods is different (more raw!) than the notice shown in Stripe credit card tab. I think I mentioned this in a previous review, it can probably be handled as a follow up issue.

Good catch! That should be fixed in d3f5cd98285cbbcedb8f90607f2ab4b11e18d406. Even though we had a user-friendly error message stored in `stripe.error`, we were not throwing it.

> In the editor, the error notices aren't always visible, because of the tabbed UI, and the fact that it's static (editor can't interact with it). This is definitely a quirk, and limits the usefulness of generating an error notice; probably it's most important that it's displayed in the editor (if anywhere). To reproduce: break a payment method that is not the first tab (probably other ways to reproduce).

Good point. This is also happening in `main`, so I would detach it from this PR.

> Also in the editor, the cheque error notice is simply not shown at all. This is curious – and suggests that the mechanism for showing error notices has a loophole somewhere. Again, this is a contrived situation (throw in canMakePayment), so might be limited impact.

Correct, that's because in the editor we were only showing an error if `content` failed to render. But any error thrown inside `canMakePayment` would be ignored.

Considering all the above, I decided to completely change how this is implemented. It shouldn't affect the API, but it changes a bit how things work:

- Failing payment methods are never registered. Not for the admin (as previous commits were attempting) neither in the editor (as it is currently in `main`). This way, the block renders the same in the editor and the frontend.
- If a payment method fails to load, we add a notice in the editor + in the frontend when the user is an admin. Those notices are shown in the `StoreNoticesProvider` of the express/standard payment methods section.

Some benefits of this approach:

- There are no differences between frontend and editor. What an admin sees in the frontend is exactly the same they see in the editor.
- If a payment method throws in `canMakePayment`, that error is now shown to admins/editor. Before, it was missed and only errors in the component were shown.
- If more than one payment method fails, the errors stack up.
- The only difference between admins/editor compared to what a regular user sees are the error notices. The payment methods that failed to register silently fail for shoppers.
- I think it also accomplishes everything you listed in `When I'm testing, I keep these principles in mind:`.

haszari reviewed (approved):
Excellent job @Aljullu !! 🙌 

The new approach seems really robust to me – should work well as more payment methods add support for blocks. I'm curious too whether the new error-catching support in this PR will help us fix various edge-cases for payment errors (e.g. #2743). 

I tested a variety of scenarios with admin, editor, and shopper roles, saved cards, various payment methods broken in different ways, and saved card option enabled/disabled, and didn't see any issues.

# Ship it!!! 🎉 🚢 :shipit:

Aljullu commented on assets/js/base/components/payment-methods/express-payment/checkout-express-payment.js:
```diff
@@ -3,22 +3,34 @@
+		// when a payment method fails to register.
+		if ( isEditor || CURRENT_USER_IS_ADMIN ) {
+			return (
+				<StoreNoticesProvider context="wc/express-payment-area"></StoreNoticesProvider>
```
Absolutely! Changed in 3d5309efc1eb53ea58bd3922e9e536a74b5a9d70.

Aljullu commented on assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js:
```diff
@@ -57,7 +42,8 @@ const handleRegistrationError = ( error ) => {
 	registeredPaymentMethods,
-	paymentMethodsSortOrder
+	paymentMethodsSortOrder,
+	noticeContext
```
Added in 4d10aaca7dfacb609a16be780a0fc5986549cf4c.

Aljullu commented on assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js:
```diff
@@ -118,10 +99,25 @@ const usePaymentMethodRegistration = (
 						throw new Error( canPay.error.message );
 					}
 					addAvailablePaymentMethod( paymentMethod );
+					removeNotice(
```
Good point... The notice is only added on registration, so probably no need for the `removeNotice`. Removed this line in 93d37bc80ceea183e84a7b676ba8016d2a7a19ff.

Aljullu commented on assets/js/payment-method-extensions/payment-methods/stripe/payment-request/index.js:
```diff
@@ -32,6 +32,9 @@ function paymentRequestAvailable( currencyCode ) {
 			isStripeInitialized = true;
 			return canPay;
 		}
+		if ( stripe.error && stripe.error instanceof Error ) {
```
I'm not sure about this one... Considering in the next line we call `throw stripe.error`, we need to ensure it's an instance of `Error`.

We could do something like:

```JS
if ( stripe.error ) {
	if ( stripe.error instanceof Error ) {
		throw stripe.error;
	} else {
		throw new Error( stripe.error );
	}
}
```

But I would like to ensure there is a case where that's needed before adding the extra logic.

Aljullu commented:
Thanks for all these reviews and helping me shape this up, @haszari. Really appreciated!

> I'm curious too whether the new error-catching support in this PR will help us fix various edge-cases for payment errors (e.g. #2743).

I didn't take a deep look at that issue. I suppose if the error is in `canMakePayment` or in the components, it might be easier to debug once this PR is merged, but if the error is thrown somewhere else, this PR might not make any difference.
